### PR TITLE
fix: `no-missing-label-refs` rule does not respect escaping

### PR DESCRIPTION
**Pull request by lumirlumir**
#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request?

Hello, 

I've fixed https://github.com/eslint/markdown/issues/346.

#### What changes did you make? (Give an overview)

I added negative lookbehind `(?<!\\)` to the regex to properly escape `[` and `]`, and also added some tests.

image

#### Related Issues

fix: #346

### CODE REVIEW

File: tests/rules/no-missing-label-refs.test.js
Comment (fasttime) [inline tests/rules/no-missing-label-refs.test.js:1]: Can we add an invalid test for `"\\[[foo]\\]"` since this is also working as intended now?

Comment (lumirlumir) [reply tests/rules/no-missing-label-refs.test.js:1]: Sure! I've added an invalid test in https://github.com/eslint/markdown/pull/348/commits/57f9d26d864eb0f0e77c33cb97a1fd1819bf020c.

**Review by nzakas — DISMISSED** [review]
LGTM. Would like @fasttime to verify his feedback before merging.

File: src/rules/no-missing-label-refs.js [hunk 1]
```diff
@@ -40,7 +40,8 @@ function findMissingReferences(node, nodeText) {
 	 * `left` is the content between the first brackets. It can be empty.
 	 * `right` is the content between the second brackets. It can be empty, and it can be undefined.
 	 */
-	const labelPattern = /\[(?<left>[^\]]*)\](?:\[(?<right>[^\]]*)\])?/dgu;
+	const labelPattern =
+		/(?<!\\)\[(?<left>[^\]]*)(?<!\\)\](?<!\\)(?:\[(?<right>[^\]]*)(?<!\\)\])?/dgu;
```
Comment (fasttime) [inline src/rules/no-missing-label-refs.js:44]: This is still not working well when the label itself contains escaped brackets, for example `"[\\[foo\\]]"` should trigger a violation, but it doesn't. I think we could fix that by improving the subpatterns where labels are matched. Maybe like this, or similar:

```suggestion
	const labelPattern =
		/(?<!\\)\[(?<left>(?:\\.|[^\]])*)(?<!\\)\](?<!\\)(?:\[(?<right>(?:\\.|[^\]])*)(?<!\\)\])?/dgu;
```

Comment (lumirlumir) [reply src/rules/no-missing-label-refs.js]: @fasttime 

Thanks! I shoud have looked more closely at the edge cases.

I've added more test cases and updated the regex in https://github.com/eslint/markdown/pull/348/commits/2f2dc47a7f89e8c55e83194399c10e262c8f88cc.

**Review by fasttime — APPROVED** [review]
LGTM, thanks!
